### PR TITLE
Autogenerate sctool docs from yaml files

### PR DESCRIPTION
**Pull request by dgarcia360**
Closes https://github.com/scylladb/scylla-manager/issues/2930

Autogenerates ``sctool`` docs from YAML files using the extension ``sphinxcontrib.datatemplates``.

## How to test this PR

1. Clone this PR. For more information, see [Cloning pull requests locally](https://docs.github.com/en/github/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally).

2. Enter the docs folder, and run:

```
make preview
````

3. Open http://127.0.0.1:5500/sctool/ with your favorite browser and preview the docs.

## Before merging this PR ...

Would it be possible to update the autogenerated YAML files? 

1. Sphinx needs a special character to detect lists. For example:

```
    '*' - matches any number of any characters including none
    '?' - matches any single character
    '[abc]' - matches one character given in the bracket
    '[a-z]' - matches one character from the range given in the bracket
```

Must be:

```
    * '*' - matches any number of any characters including none
    * '?' - matches any single character
    * '[abc]' - matches one character given in the bracket
    * '[a-z]' - matches one character from the range given in the bracket
```

2. To highlight inline-code (e.g. ``abc``), replace the single quotes as follows ``'abc'`` -> ` ``[abc]`` `.

## Future improvements

* [Add a collapsible component](https://github.com/scylladb/sphinx-scylladb-theme/issues/276) in the theme to hide "inherited options".
* Support for more than one example per command. For example, instead of:

```
example: |
  sctool status -c prod-cluster
  Datacenter: eu-west
```

The example key could look like this:

```
examples: 
   - name: Example A 
      description: Example description
     command: sctool xyz
   - name: Example B
      description: Example description
     command: sctool xyz
```

Adding the new keys ``name`` and ``description`` will also be possible to format commands as code-blocks and descriptions as regular text.

### CODE REVIEW

File: docs/source/sctool/partials/sctool_backup_delete.yaml [hunk 1]
```diff
@@ -25,7 +25,7 @@ options:
   shorthand: T
   default_value: '[]'
   usage: |
-    A list of snapshot `tag`s sparated by a comma.
+    A list of snapshot `tags` sparated by a comma.
```
Comment (dgarcia360) [inline docs/source/sctool/partials/sctool_backup_delete.yaml:28]: @mmatczuk Could you please update the code in `master` to generate this output?

File: docs/source/sctool/partials/sctool_backup_list.yaml [hunk 2]
```diff
@@ -45,11 +45,11 @@ options:
     Specifies maximal snapshot `date` expressed in RFC3339 form or 'now[+duration]', ex. 'now+3d2h10m'.
     Valid units are:
 
-    'd' - days
-    'h' - hours
-    'm' - minutes
-    's' - seconds
-    'now' - happens immediately
+    * ``d`` - days
```
Comment (dgarcia360) [inline docs/source/sctool/partials/sctool_backup_list.yaml:48]: @mmatczuk Could you please  update the code in `master` to generate this output (applies to all lists)? Sphinx needs the character ``*`` to format it as a list.

File: docs/source/sctool/partials/sctool_backup_list.yaml [hunk 3]
```diff
@@ -45,11 +45,11 @@ options:
...
+    * ``h`` - hours
+    * ``m`` - minutes
+    * ``s`` - seconds
+    * ``now`` - happens immediately
 - name: min-date
   usage: |
     Specifies minimal snapshot `date` expressed in RFC3339 form or 'now[+duration]', ex. 'now+3d2h10m'.
```
Comment (dgarcia360) [inline docs/source/sctool/partials/sctool_backup_list.yaml:55]:
```suggestion
    Specifies minimal snapshot ``date`` expressed in RFC3339 form or ``now[+duration]``, ex. ``now+3d2h10m``.
Write
```

@mmatczuk Could you please update the code in ``master`` to generate this output (applies to all inline code blocks). The current documentation is using three different forms (', `` and `).

File: docs/source/sctool/backup.rst [hunk 4]
```diff
@@ -476,8 +121,6 @@ backup files parameters
```
Comment (dgarcia360) [inline docs/source/sctool/backup.rst:121]: @mmatczuk I couldn't find a YAML file for the command ``backup files``. Should we delete it from the doc (rst)?

File: docs/source/sctool/backup.rst [hunk 5]
```diff
@@ -35,175 +35,8 @@ You cannot initiate a backup without a cluster. Make sure you add a cluster (:re
 backup
 ======
```
Comment (dgarcia360) [inline docs/source/sctool/backup.rst:37]: @mmatczuk The command ``sctool_backup_validate_update.yaml`` is not in the docs (rst) file. Should I create a new heading and include it?

File: docs/source/sctool/task.rst [hunk 6]
```diff
@@ -17,21 +17,23 @@ The task command set allows you to schedule, start, stop and modify tasks.
```
Comment (dgarcia360) [inline docs/source/sctool/task.rst:17]: @mmatczuk According to the YAML files, now  ``sctool task <command> [flags] [global flags]`` is ``sctool <command> [flags] [global flags]``, right?

File: docs/source/sctool/task.rst [hunk 7]
```diff
@@ -17,21 +17,23 @@ The task command set allows you to schedule, start, stop and modify tasks.
...
+   * - :ref`task progress<task-progress>`
      - Show the task progress.
-   * - `task start`_
+   * - :ref`task start<task-start>`
      - Start executing a task.
-   * - `task stop`_
+   * - :ref`task stop<task-stop>`
      - Stop executing a task.
-   * - `task update`_
+   * - :ref`task update<task-update>`
```
Comment (dgarcia360) [inline docs/source/sctool/task.rst:32]: TODO @dgarcia360 : Update this list once I have more clear which commands remain.

File: docs/source/sctool/task.rst [hunk 8]
```diff
@@ -17,21 +17,23 @@ The task command set allows you to schedule, start, stop and modify tasks.
...
      - Modify a task.
 
+.. _task-delete:
+
 task delete
```
Comment (dgarcia360) [inline docs/source/sctool/task.rst:37]: @mmatczuk I couldn't find an equivalent command for ``task delete``. Should we delete it from the docs (rst)?

File: docs/source/sctool/task.rst [hunk 9]
```diff
@@ -71,6 +65,8 @@ You will have to create a new one.
    sctool task delete -c prod-cluster repair/143d160f-e53c-4890-a9e7-149561376cfd
 
 
+.. _task-history:
+
 task history
```
Comment (dgarcia360) [inline docs/source/sctool/task.rst:70]: @mmatczuk I couldn't find an equivalent command for ``task history``. Should we delete it from the docs (rst)?

File: docs/source/sctool/task.rst [hunk 10]
```diff
@@ -486,27 +349,8 @@ This can be retrieved using the command ``sctool task list``.
...
-
-=====
-
-.. include:: ../_common/param-cluster.rst
-
-=====
+.. datatemplate:yaml:: partials/sctool_stop.yaml
+   :template: command.tmpl
 
 Example: task stop
```
Comment (dgarcia360) [inline docs/source/sctool/task.rst:355]: @mmatczuk I couldn't find an equivalent command for ``task stop``. Should we delete it from the docs (rst)?

File: docs/source/sctool/task.rst [hunk 11]
```diff
@@ -71,6 +65,8 @@ You will have to create a new one.
    sctool task delete -c prod-cluster repair/143d160f-e53c-4890-a9e7-149561376cfd
```
Comment (dgarcia360) [inline docs/source/sctool/task.rst:65]: @mmatczuk If the command `sctool task <command>` was renamed to `sctool <command>`, we should adapt the code examples manually. 

For the task related commands: Should we keep the examples as they are and update them in a second iteration? Or, would you prefer to delete them for now?

File: docs/source/sctool/task.rst [hunk 12]
```diff
@@ -133,77 +121,8 @@ Example: task history
 task list
```
Comment (dgarcia360) [inline docs/source/sctool/task.rst:121]: TODO @dgarcia360 : Update headings to match the new command names.

File: docs/source/sctool/cluster.rst [hunk 13]
```diff
@@ -80,32 +60,13 @@ Example (IPv6):
```
Comment (dgarcia360) [inline docs/source/sctool/cluster.rst:60]: @dgarcia360 TODO: Some YAML files have a property "example". I'll try to render the examples also from this new key.

File: docs/source/sctool/task.rst (hunk 6 above)
Comment (mmatczuk) [reply docs/source/sctool/task.rst:17]: Yes.

File: docs/source/sctool/task.rst (hunk 11 above)
Comment (mmatczuk) [reply docs/source/sctool/task.rst:65]: There are examples in some of the input files. Use those, others remove for now.

**Comment by mmatczuk** [comment]
@dgarcia360 please note that there are changes in commands and command layout.
There is no point to find an exact mapping here.

File: docs/source/sctool/partials/sctool_progress.yaml [hunk 14]
```diff
@@ -50,11 +50,13 @@ options:
 - name: run
   default_value: latest
   usage: |
-    Show progress of a particular run, see sctool info to get the `ID`s.
+    Show progress of a particular run, see sctool info to get the `IDs`.
```
Comment (dgarcia360) [inline docs/source/sctool/partials/sctool_progress.yaml:53]: @mmatczuk Could you please update the code in `master` to generate this output? Sphinx complains if the `s` is after the "`" symbol.

**Comment by mmatczuk** [comment]
@dgarcia360 can we merge it i.e. won't it break the 2.5 2.6 and so on?
We want that for master only for now.

**Comment by dgarcia360** [comment]
@mmatczuk I've tested it locally and it should not break previous versions.

**Comment by dgarcia360** [comment]
**Notes from 05 Jan 2022 call with @tzach** 

Once we merge this PR:

- [ ] @dgarcia360 will update the docstrings as described in  **Before merging this PR**, and submit a PR.
- [ ] @dgarcia360  will create issues for the recommendations described in **Future improvements**.

**Pull request merged**